src/engraving/rendering/dev/accidentalslayout.cpp

```diff
@@ -0,0 +1,1112 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2023 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include <cfloat>
+
+#include "accidentalslayout.h"
+#include "layoutcontext.h"
+#include "../paddingtable.h"
+#include "shape.h"
+#include "tlayout.h"
+
+#include "dom/accidental.h"
+#include "dom/chord.h"
+#include "dom/ledgerline.h"
+#include "dom/note.h"
+#include "dom/score.h"
+
+using namespace mu::engraving;
+using namespace mu::engraving::rendering::dev;
+
+void AccidentalsLayout::layoutAccidentals(const std::vector<Chord*>& chords, LayoutContext& ctx)
+{
+    std::vector<Accidental*> allAccidentals = collectAccidentals(chords);
+
+    if (allAccidentals.empty()) {
+        return;
+    }
+
+    for (Accidental* acc : allAccidentals) {
+        acc->computeMag();
+        TLayout::layoutAccidental(acc, acc->mutldata(), ctx.conf());
+    }
+
+    AccidentalPlacementEngine accidentalPlacementEngine(allAccidentals, chords);
+    accidentalPlacementEngine.doAccidentalPlacement();
+}
+
+std::vector<Accidental*> AccidentalsLayout::collectAccidentals(const std::vector<Chord*> chords)
+{
+    std::vector<Accidental*> accidentals;
+
+    for (Chord* chord : chords) {
+        for (Note* note : chord->notes()) {
+            Accidental* acc = note->accidental();
+            if (!acc) {
+                continue;
+            }
+            acc->setPos(0.0, 0.0);
+            bool isRedundant = false;
+            for (Accidental* otherAcc : accidentals) {
+                if (otherAcc->accidentalType() == acc->accidentalType() && otherAcc->line() == acc->line()) {
+                    isRedundant = true;
+                    break;
+                }
+            }
+            if (isRedundant) {
+                acc->setbbox(RectF());
+                continue;
+            }
+            accidentals.push_back(acc);
+        }
+    }
+
+    return accidentals;
+}
+
+AccidentalsLayout::AccidentalPlacementEngine::AccidentalPlacementEngine(const std::vector<Accidental*>& allAccidentals,
```

Comment:
I don't really like this class and the fact that it has members and is named as “engine”... It would be better to re-implement it as a bunch of static (private) methods of AccidentalsLayout. You can add a private struct that will store all these members and pass it as an argument to the static methods